File: examples/01-hello-agent/agent.py
"""thaghr's minimal example agent.

Not a demonstration of a real production agent, just enough surface
for the campaign runner to exercise: one call out to an LLM provider,
returning what happened. thaghr's runner calls `run()` once per trial,
handing it an httpx.Client whose transport may or may not be injecting
faults.

Defaults to OpenAI. Works against any OpenAI-compatible endpoint (Gemini,
for instance) by setting THAGHR_DEMO_BASE_URL and THAGHR_DEMO_MODEL, so
you don't need an OpenAI key just to run this example. For Gemini:
OPENAI_API_KEY=<your gemini key>
THAGHR_DEMO_BASE_URL=https://generativelanguage.googleapis.com/v1beta/openai/
THAGHR_DEMO_MODEL=gemini-3.6-flash (check ai.google.dev/gemini-api/docs/openai
for the current model name, Google updates these)
"""
from __future__ import annotations

import os

import httpx
import openai


def run(http_client: httpx.Client | None = None) -> dict:
    """Called once per trial. Returns a dict describing the outcome.

    Raises whatever the OpenAI SDK raises (RateLimitError, APIConnectionError,
    etc.) on failure; the campaign runner is responsible for catching that,
    not this function.
    """
    client = openai.OpenAI(
        api_key=os.environ.get("OPENAI_API_KEY", "thaghr-demo-key"),
        base_url=os.environ.get("THAGHR_DEMO_BASE_URL"),
        http_client=http_client,
        max_retries=2,
    )
    model = os.environ.get("THAGHR_DEMO_MODEL", "gpt-4o-mini")
    completion = client.chat.completions.create(
        model=model,
        messages=[{"role": "user", "content": "Say hello in exactly three words."}],
    )
    usage = completion.usage.model_dump() if completion.usage else {}
    return {
        "content": completion.choices[0].message.content,
        "prompt_tokens": usage.get("prompt_tokens", 0),
        "completion_tokens": usage.get("completion_tokens", 0),
    }

File: examples/06-node-proxy/agent.mjs
// thaghr Phase 6 example: the OpenAI JS SDK gets the same fault stack as
// the Python httpx-transport integration (Phase 2), with the only
// client-side change being `baseURL`.
//
// Run against:
//   1. `thaghr proxy` fronting a real OpenAI-compatible endpoint, or
//   2. `thaghr proxy` fronting the local fake_upstream.py in this
//      directory, for a deterministic, zero-cost demo.
//
// Env vars (all optional, mirrors examples/01-hello-agent/agent.py):
//   THAGHR_PROXY_URL   base URL of the running `thaghr proxy` (default http://127.0.0.1:8135/v1)
//   OPENAI_API_KEY     forwarded to the SDK; the proxy doesn't check it,
//                      the real upstream does (default a dummy key, fine
//                      against fake_upstream.py)
//   THAGHR_DEMO_MODEL  model name to request (default gpt-4o-mini)

import OpenAI from "openai";

const baseURL = process.env.THAGHR_PROXY_URL ?? "http://127.0.0.1:8135/v1";
const model = process.env.THAGHR_DEMO_MODEL ?? "gpt-4o-mini";

const client = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY ?? "thaghr-demo-key",
  baseURL,
  maxRetries: 2,
});

async function run() {
  try {
    const completion = await client.chat.completions.create({
      model,
      messages: [{ role: "user", content: "Say hello in exactly three words." }],
    });
    console.log("thaghr proxy demo: call succeeded");
    console.log("  content:", completion.choices[0].message.content);
    console.log("  usage:", completion.usage);
  } catch (err) {
    // A faulted call surfaces here as whatever error the OpenAI JS SDK
    // raises for the injected status code (RateLimitError for 429), the
    // same shape a real production 429 would take. This is the point of
    // Phase 6: the SDK's own retry/error handling engages against an
    // injected fault exactly as it would against a real one, and nothing
    // about the client code above changed to make that happen.
    console.log("thaghr proxy demo: call failed");
    console.log("  name:", err.name);
    console.log("  status:", err.status);
    console.log("  message:", err.message);
    process.exitCode = 1;
  }
}

run();
